##  1. Python Program to Check Prime Number

In [2]:
#Using FOR Loop

def check_prime(n):
    flag=0
    if n>1:
        for i in range(2,n//2+1):
            if n%i==0:
                flag=1
                break
    else:
        flag=-1
        

    if flag==1:
        print("No, it is not a prime no!")
    elif flag==0:
        print("Yes, it is a prime no!")
    else:
        print("Invalid no!")

In [5]:
check_prime(15)
check_prime(-5)
check_prime(88)

No, it is not a prime no!
Invalid no!
No, it is not a prime no!


In [11]:
# Using a for...else

def chk_prime2(n):
    if n>1:
        for i in range(2,n//2+1):
            if n%i==0:
                print("Not a prime no!")
                break
        else:
            print("It is a prime no!")
    else:
        print("Not a prime no!") 



In [13]:
chk_prime2(41)
chk_prime2(34)

It is a prime no!
Not a prime no!


## 2. Python Program to Find Prime Numbers in a Given Range

In [21]:
def range_prime(l,u):
    cnt=0
    for i in range(l,u):
        if n>1:
                for j in range(2,i//2+1):
                    if i%j==0:
                        break
                else:
                    print(i, end =' ')
                    cnt+=1
    print("\nTotal count between range is: ", cnt)

In [22]:
range_prime(4,56)

5 7 11 13 17 19 23 29 31 37 41 43 47 53 
Total count between range is:  14


In [25]:
# Print all no within a range and also give count!

def prime_cnt():
    l= int(input("Enter lower range: "))
    u= int(input("Enter upper range: "))
    cnt=1

    for i in range(l,u):
        if i>1:
            for j in range(2,i):
                if (i%j==0):
                    break
            else:
                cnt+=1
                print(i, end=' ')

    print("\nTotal count between lower and upper range is: ", cnt-1)
    


In [26]:
prime_cnt()

Enter lower range: 120
Enter upper range: 150
127 131 137 139 149 
Total count between lower and upper range is:  5


## 3. Python Program to Check whether a Number is Prime or Not using Recursion
 

In [1]:
def rec_prime(n,i=2):
    if n==i:
        return True
    elif n%i==0:
        return False
    return rec_prime(n,i+1)

In [2]:
n = 4
if rec_prime(n):
    print("Yes,", n, "is Prime")
else:
    print("No,", n, "is not a Prime")

No, 4 is not a Prime


## 4. Python Program to Check Whether a Given Number is Perfect Number

Case 1:
Enter any number: 6
The number is a Perfect number!


When a number is equal to the sum of all of its positive divisors except itself, it is said to be a Perfect Number.

For example:

n = 6

Positive divisors of 6 except itself are 1, 2 and 3

Sum of divisors is 1 + 2 + 3 = 6

Hence 6 is a perfect number.



In [7]:
#Using For Loop

def chk_pernum(n):
    total=0
    for i in range(1,n):
        if (n%i==0):
            total+=i
            
    if total==n:
        print("Yes, it is a perfect no!")
    else:
        print("No, it is not a perfect no!")

In [28]:
chk_pernum(6)
chk_pernum(25)
chk_pernum(28)

Yes, it is a perfect no!
No, it is not a perfect no!
Yes, it is a perfect no!


In [3]:
#Using While Loop

def chk_pernum2(n):
    total=0
    i=1
    
    while i<n:
        if (n%i==0):
            total+=i
        i+=1
            
    if total==n:
        print("Yes, it is a perfect no!")
    else:
        print("No, it is not a perfect no!")

In [6]:
chk_pernum2(6)
chk_pernum2(25)
chk_pernum2(28)

Yes, it is a perfect no!
No, it is not a perfect no!
Yes, it is a perfect no!


## 5. Python Program to Check Armstrong Number

A number of n-digits can be defined as an Armstrong number if and only if the sum of its digits powered to n is equal to the number itself. For example, 1634 is an Armstrong number as

1634= 1^4+ 6^4 + 3^4+ 4^4

--------------------------------
Input : 153
Output : Yes
153 is an Armstrong number.
1*1*1 + 5*5*5 + 3*3*3 = 153


In [10]:
#Using While Loop & Len function

def chk_armstrong(num):
    total=0
    digit=0
    temp=num
    l=len(str(temp))
    
    while temp>0:
        digit=temp%10
        total+=digit**l
        temp//=10
        
    if num==total:
        print("Yes, it is a armstrong no!")
    else:
        print("No, it is not a armstrong no!")

In [14]:
chk_armstrong(1634)
chk_armstrong(153)
chk_armstrong(123)

Yes, it is a armstrong no!
Yes, it is a armstrong no!
No, it is not a armstrong no!


In [2]:
def count_digit(num):
    cnt=0
    while num>0:
        num//=10
        cnt+=1
    return cnt

In [20]:
print("No of digit in given no: 1634 is",count_digit(1634))
print("No of digit in given no: 153 is",count_digit(153))
print("No of digit in given no: 123 is",count_digit(123))
print("No of digit in given no: 2 is",count_digit(2))
print("No of digit in given no: 99999 is",count_digit(99999))

No of digit in given no: 1634 is 4
No of digit in given no: 153 is 3
No of digit in given no: 123 is 3
No of digit in given no: 2 is 1
No of digit in given no: 99999 is 5


In [24]:
#Using While Loop & digit count function

def chk_armstrong2(num):
    total=0
    digit=0
    temp=num
    l=count_digit(temp)
    
    while temp>0:
        digit=temp%10
        total+=pow(digit,l)
        temp//=10
        
    if num==total:
        print("Yes, it is a armstrong no!")
    else:
        print("No, it is not a armstrong no!")

In [26]:
chk_armstrong2(1634)
chk_armstrong2(153)
chk_armstrong2(123)
chk_armstrong2(407)

Yes, it is a armstrong no!
Yes, it is a armstrong no!
No, it is not a armstrong no!
Yes, it is a armstrong no!


In [28]:
def check_armstrong(num,n1,sum,temp):
    if temp==0:
        if sum==num:
            return True
        else:
            return False
    digit = temp % 10
    sum = sum + digit**n1
    temp = temp//10
    return check_armstrong(num, n1, sum, temp)

num = int(input("Enter a number: "))
sum = 0
n1 = len(str(num))
temp = num
res = check_armstrong(num,n1,sum,temp)

if res:
    print(num,"is an Armstrong number")
else:
    print(num,"is not an Armstrong number")

Enter a number: 153
153 is an Armstrong number


## 6. Python Program to Find Armstrong Number between an Interval

In [13]:
def print_armstrong():

    l = int(input("Enter lower range: "))  
    u = int(input("Enter upper range: "))  
    for num in range(l,u+1):
        
        tot=0
        temp=num
        l=count_digit(temp)

        while temp>0:
            digit=temp%10
            tot+=digit**l
            temp//=10
        
        if num==tot:
            print(num)
    

In [14]:
print_armstrong()

Enter lower range: 100
Enter upper range: 500
153
370
371
407


## 7. Sum of First N Natural Numbers in Python

Example
Input : num = 8
Output : 36

Where first 8 number is 1, 2, 3, 4, 5, 6, 7, 8
Sum of numbers = 1 + 2 + 3 + 4 + 5 + 6 + 7 + 8 = 36

In [9]:
tot=([i for i in range(8+1)])
sum(tot)

36

In [10]:
#Using FOR Loop

n=8
tot=0
for i in range(n+1):
    tot+=i
tot

36

In [5]:
#USing recursion

def rec_nsum(n):
    if n==0:
        return 0
    else:
        return n+rec_nsum(n-1)

In [6]:
rec_nsum(8)

36

## 8. Python Program to Check if a Number is a Strong Number

Strong Number
A Number that is equal to the sum of the factorial of it's individual digits is known as Strong Number.

Example
Input : 145
Output : It's a Strong Number.
Explanation : Number = 145
145 = 1! + 4! + 5!
145 = 1 + 24 + 120

In [11]:
def fact(n):
    if n<=1:
        return 1
    else:
        return n*fact(n-1)

In [13]:
fact(5)

120

In [19]:
#Using WHILE loop and using Fact program in it!

def chk_strong(num):
    temp=num
    tot=0
    
    while temp>0:
        digit=temp%10
        tot+=fact(digit)
        temp//=10
        
    if tot==num:
        print(num,"is a Strong no!")
    else:
        print(num,"is not a Strong no!")

In [22]:
chk_strong(145)
chk_strong(154)
chk_strong(30)

145 is a Strong no!
154 is not a Strong no!
30 is not a Strong no!


## 9. Python Program to Print Numbers in a Range Without using Loops

In [34]:
def print_num(n):
    if n>0:
        print_num(n-1)
        print(n, end=' ')
        

In [35]:
print_num(8)

1 2 3 4 5 6 7 8 

In [42]:
def prnt_n(l,u):    
    if (u>l):
        print(l, end=' ')
        prnt_n(l+1,u)
        

In [43]:
prnt_n(2,7)

2 3 4 5 6 

## 10. Python Program to Print the Natural Numbers Summation Pattern

Case 1:
Enter a number: 4
1 = 1
1 + 2 = 3
1 + 2 + 3 = 6
1 + 2 + 3 + 4 = 10

In [46]:
number=4
for j in range(1, number+1):
    a = []
 
    for i in range(1, j+1):
        print(i, sep=" ", end=" ")
        if(i < j):
            print("+", sep=" ", end=" ")
        a.append(i)
        
    print("=", sum(a))
print()

1 = 1
1 + 2 = 3
1 + 2 + 3 = 6
1 + 2 + 3 + 4 = 10



In [57]:
def print_natural_sum(n):
    last = "1"
    total = 1
    
    #print for 1
    print(last, "=", total)
    for i in range(2, n+1):
        
        total += i
        # update the string and print it
        last += " + " + str(i)
        print(last, "=", total)
        

In [64]:
print_natural_sum(5)

1 = 1
1 + 2 = 3
1 + 2 + 3 = 6
1 + 2 + 3 + 4 = 10
1 + 2 + 3 + 4 + 5 = 15


## 11. Python Program to Find the Prime Factors of a Number

In [19]:
#Written prime no check program so that we can use it later

def chk_prime(n):
    if n>1:
        for i in range(2,n//2+1):
            if(n%i)==0:
                return False
                break
        else:
            return True

In [25]:
#Using FOR loop and chk_prime program

def prime_fac(n):
    for i in range(1,n):
        if (n%i) ==0 :
            if chk_prime(i):
                print(i, end=' ')

In [28]:
prime_fac(110)

2 5 11 

In [29]:
prime_fac(25)

5 

In [30]:
prime_fac(200)

2 5 

In [33]:
# Python Program to find Prime Factors of a Number
 
Number = int(input(" Please Enter any Number: "))

for i in range(2, Number + 1):
    if(Number % i == 0):
        isprime = 1
        for j in range(2, (i //2 + 1)):
            if(i % j == 0):
                isprime = 0
                break
            
        if (isprime == 1):
            print(" %d is a Prime Factor of a Given Number %d" %(i, Number))

 Please Enter any Number: 110
 2 is a Prime Factor of a Given Number 110
 5 is a Prime Factor of a Given Number 110
 11 is a Prime Factor of a Given Number 110


## 12. Python Program to Check If Two Numbers are Amicable Numbers or Not

In [1]:
def sum_divsor(n):
    sum_n=0
    
    for i in range(1,n):
        if (n%i)==0:
            sum_n+=i
    return sum_n
    

In [62]:
sum_divsor(284)

220

In [65]:
def chk_amicable(n1,n2):
    
    if sum_divsor(n1)==n2 and sum_divsor(n2)==n1:
        print("Yes! both",n1 ,"&" ,n2 , "are amicable")
    else:
        print("No! both",n1 ,"&" ,n2 , "are not amicable")
    

In [66]:
chk_amicable(220,284)
chk_amicable(60,84)
chk_amicable(1210,1184)

Yes! both 220 & 284 are amicable
No! both 60 & 84 are not amicable
Yes! both 1210 & 1184 are amicable


In [6]:
def chk_amicable(n1,n2):
    
    if sum_divsor(n1)==n2 and sum_divsor(n2)==n1:
        print("Yes!! both {} & {} are amicable".format(n1,n2))
    else:
        print('No!! both {} & {} are not amicable'.format(n1,n2))
    

In [7]:
chk_amicable(220,284)
chk_amicable(60,84)
chk_amicable(1210,1184)

Yes!! both 220 & 284 are amicable
No!! both 60 & 84 are not amicable
Yes!! both 1210 & 1184 are amicable


## 13. Python Program to Find Whether a Number is a Power of Two

In [11]:
def power_two(n):
    mul=2
    
    while mul<n-1:
        mul*=2
    
    if (mul==n):
        print(n, "It is a power of 2!")
    else:
        print(n, "No it is not a power of 2!")

In [100]:
power_two(8)
power_two(9)
power_two(16)
power_two(55)

8 It is a power of 2!
9 No it is not a power of 2!
16 It is a power of 2!
55 No it is not a power of 2!


In [15]:
#Using While Loop

def isPowerOfTwo(n):
    if (n == 0):
        return False
    while (n != 1):
        if (n % 2 != 0):
            return False
        n = n // 2
             
    return True

In [16]:
# Driver code
if(isPowerOfTwo(31)):
    print('Yes')
else:
    print('No')

No


In [21]:
def is_power_of_two(n):
    return (n != 0) and (n & (n-1) == 0)

In [22]:
is_power_of_two(8)

True

In [23]:
is_power_of_two(9)

False

## 14. Python Program to Calculate the Power using Recursion

In [42]:
def calc_pow(num,power):
    if power==0:
        return 1
    else:
        return num*calc_pow(num,power-1)

In [43]:
calc_pow(3,4)

81

In [44]:
calc_pow(2,5)

32

## 15. Python Program to Find Product of Two Numbers using Recursion

Case 1:
Enter first number: 12
Enter second number: 10
Product is:  120

In [1]:
def rec_product(a,b):
    if a<b:
        return rec_product(b,a)
    elif(b!=0):
        return (a+rec_product(a,b-1))
    else:
        return 0
    

In [3]:
rec_product(4,11)

44

In [2]:
rec_product(4,16)

64

In [49]:
#Product without using multiplication

def product(a,b):
    pro=0
    x=max(a,b)
    y=min(a,b)
    for i in range(x):
        pro+=y
    print("Product of {} & {} are {}".format(a,b,pro))
    

In [51]:
product(4,14)
product(12,10)

Product of 4 & 14 are 56
Product of 12 & 10 are 120


## 16. Python Program to Find All Perfect Squares in the Given Range

A number is said to be a perfect square number if it can be obtained by the multiplication of two equal integer numbers.

 

for example,

16 is a perfect square number because it can be obtained by the multiplication of two equal integers (4*4=16).

In [29]:
def print_persqr():
    l=int(input("Enter Lower Range: "))
    u=int(input("Enter Upper Range: "))
    
    for i in range(l,u):
        if i*i>=u:
            break
        print(i*i, end=' ')

In [30]:
print_persqr()

Enter Lower Range: 1
Enter Upper Range: 127
1 4 9 16 25 36 49 64 81 100 121 

In [31]:
print_persqr()

Enter Lower Range: 1
Enter Upper Range: 30
1 4 9 16 25 

In [32]:
#Checking Perfect Square No!

from math import sqrt

def isPerfectSquare(x):
    if x >= 0:
        sr = int(sqrt(x))
        return (sr * sr) == x
    return False


In [33]:
n = 84
if isPerfectSquare(n):
    print("True")
else:
    print("False")

False


In [34]:
#Other Method

def squares(a, b):
    lists=[]
    # Traverse through all numbers
    for i in range (a,b+1):
        j = 1;
        while j*j <= i:
            if j*j == i:
                 lists.append(i)  
            j = j+1
        i = i+1
    return lists
print(squares(1, 30))


[1, 4, 9, 16, 25]


In [35]:
def print_perfect_squares(start, end):
    for i in range(start, end+1):
        root = int(i**0.5)
        if root*root == i:
            print(i)

# Example usage
print_perfect_squares(1, 20)   # Output: 1 4 9 16


1
4
9
16


## 17. Python Program to Print All Possible Combinations of Three Digits

Input: [1, 2, 3]

Output:
1 2 3
1 3 2
2 1 3
2 3 1
3 1 2
3 2 1

In [36]:
def comb(L):
      
    for i in range(3):
        for j in range(3):
            for k in range(3):
                  
                # check if the indexes are not
                # same
                if (i!=j and j!=k and i!=k):
                    print(L[i], L[j], L[k])
                      
# Driver Code
comb([1, 2, 3])

1 2 3
1 3 2
2 1 3
2 3 1
3 1 2
3 2 1


In [37]:
from itertools import permutations
  
# Get all combination of [1, 2, 3]
# of length 3
comb = permutations([1, 2, 3], 3)
  
for i in comb:
    print(i)

(1, 2, 3)
(1, 3, 2)
(2, 1, 3)
(2, 3, 1)
(3, 1, 2)
(3, 2, 1)
